### DISCUSSION TITLE
Lesson 11 ModdNft is not deploying on the anvil chain

### DISCUSSION THREAD

**Discussion by NeoRusi** [Cyfrin/foundry-full-course-cu#951] [Q&A] [answered]
I am deploying the moodNft on the anvil chain and this is what I receive as an output 
```
[⠆] Compiling...
[⠔] Compiling 2 files with 0.8.19
[⠒] Solc 0.8.19 finished in 2.11s
Compiler run successful!
Traces:
  [1809601] → new DeployMoodNft@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
    └─ ← 8928 bytes of code

  [23998] DeployMoodNft::run() 
    ├─ [0] VM::readFile(./img/sad.svg) [staticcall]
    │   └─ ← "The path "./img/sad.svg" is not allowed to be accessed for read operations."
    └─ ← "The path "./img/sad.svg" is not allowed to be accessed for read operations."


Error: 
The path "./img/sad.svg" is not allowed to be accessed for read operations.
```
here is the MakeFile
```make
-include .env

.PHONY: all test clean deploy fund help install snapshot format anvil 

DEFAULT_ANVIL_KEY := 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

help:
	@echo "Usage:"
	@echo "  make deploy [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""
	@echo ""
	@echo "  make fund [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""

all: clean remove install update build

# Clean the repo
clean  :; forge clean

# Remove modules
remove :; rm -rf .gitmodules && rm -rf .git/modules/* && rm -rf lib && touch .gitmodules && git add . && git commit -m "modules"

install :; forge install Cyfrin/foundry-devops@0.0.11 --no-commit && forge install foundry-rs/forge-std@v1.5.3 --no-commit && forge install openzeppelin/openzeppelin-contracts@v4.8.3 --no-commit

# Update Dependencies
update:; forge update

build:; forge build

test :; forge test 

snapshot :; forge snapshot

format :; forge fmt

anvil :; anvil -m 'test test test test test test test test test test test junk' --steps-tracing --block-time 1

NETWORK_ARGS := --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast

ifeq ($(findstring --network sepolia,$(ARGS)),--network sepolia)
	NETWORK_ARGS := --rpc-url $(SEPOLIA_RPC_URL) --private-key $(PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
endif

deploy:
	@forge script script/DeployBasicNft.s.sol:DeployBasicNft $(NETWORK_ARGS)

mint:
	@forge script script/Interactions.s.sol:MintBasicNft $(NETWORK_ARGS)
	
deployMood:
	@forge script script/DeployMoodNft.s.sol:DeployMoodNft $(NETWORK_ARGS)

mintMoodNft:
	@forge script script/Interactions.s.sol:MintMoodNft $(NETWORK_ARGS)

flipMoodNft:
	@forge script script/Interactions.s.sol:FlipMoodNft $(NETWORK_ARGS)

```
Here is the deployMoodNft.s.sol
```solidity
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {MoodNft} from "../src/MoodNft.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";

contract DeployMoodNft is Script {
    function run() external returns (MoodNft) {
        string memory sadSvg = vm.readFile("./img/sad.svg");
        string memory happySvg = vm.readFile("./img/happy.svg");
        vm.startBroadcast();
        MoodNft moodNft = new MoodNft (
        svgToImageUri(sadSvg),
        svgToImageUri(happySvg));
        vm.stopBroadcast();
        return moodNft;
    }

    function svgToImageUri(string memory svg) public pure returns (string memory) {
        string memory baseURL = "data:image/svg+xml;base64,";
        string memory baseSvgEncoded = Base64.encode(bytes(string(abi.encodePacked(svg))));
        return string(abi.encodePacked(baseURL, baseSvgEncoded));
    }
}
```
who it can't read from the path I don't understand everything with the path is ok here is a screenshot of my paths 
<img width="183" alt="Screenshot 2023-10-01 at 23 06 14" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/7d4d4572-96c4-4850-9679-16ba77c7e5f3">

**Comment by theirrationalone** [comment, marked as answer]
You just need to add this toml snippet into your foundry.toml file
.... 

```toml
fs_permissions = [{ access = "read", path = "./img/"}]
```

**Reply by reactinary** [reply]
Thank you so much!!!

**Comment by dusandebelak** [comment]
Hi, I have the same problem:
animan@DESKTOP-3FCGQ8H:~/solidity-course/foundry-f23/second-try-NFT$ forge test
[⠢] Compiling...
[⠰] Compiling 5 files with 0.8.20
[⠔] Solc 0.8.20 finished in 3.08s
Compiler run successful!
2024-02-17T14:40:52.537814Z ERROR forge::runner: setUp failed reason=the path ../img/sad.svg is not allowed to be accessed for read operations contract=0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496

Running 1 test for test/MoodNftTest.t.sol:MoodNftTest
[FAIL. Reason: setup failed: the path ../img/sad.svg is not allowed to be accessed for read operations] setUp() (gas: 0)
Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 582.10µs

Running 1 test for test/DeployMoodNftTest.t.sol:DeployMoodNftTest
[PASS] testConvertSvgToUri() (gas: 26806)
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 2.31ms

Running 4 tests for test/BasicNftTest.t.sol:BasicNftTest
[PASS] testCanMintAndHaveABalance() (gas: 168902)
[PASS] testInitializedCorrectly() (gas: 14571)
[PASS] testMintWithScript() (gas: 431136)
[PASS] testTokenURIIsCorrect() (gas: 172203)
Test result: ok. 4 passed; 0 failed; 0 skipped; finished in 2.48ms
 
Ran 3 test suites: 5 tests passed, 1 failed, 0 skipped (6 total tests)

Failing tests:
Encountered 1 failing test in test/MoodNftTest.t.sol:MoodNftTest
[FAIL. Reason: setup failed: the path ../img/sad.svg is not allowed to be accessed for read operations] setUp() (gas: 0)

Encountered a total of 1 failing tests, 5 tests succeeded

And I have foundry.toml:
[profile.default]
src = "src"
out = "out"
libs = ["lib"]
ffi = true
fs_permissions = [
    { access = "read", path = "./img/"},
    { access = "read-write", path = "./broadcast" },
]

remappings = ['@openzeppelin/contracts=lib/openzeppelin-contracts/contracts']

And still doesn't work :(
how can I fix it?

Thank you :)

**Reply by theirrationalone** [reply]
It might be directory hierarchy lvl issue... As logs are telling that the path that is one lvl up (parent) isn't allowed for read operation. 

Try this... 
```toml
fs_permissions = [{ access = "read", path = "../img/" },]
```
Just changed the access path from `.` (current) to `.. ` (parent)

**Reply by dusandebelak** [reply]
I tried doesn't work :(

**Reply by dusandebelak** [reply]
animan@DESKTOP-3FCGQ8H:~/solidity-course/foundry-f23/second-try-NFT$ forge test
[⠢] Compiling...
No files changed, compilation skipped
2024-02-17T15:06:01.102581Z ERROR forge::runner: setUp failed reason=the path ../img/sad.svg is not allowed to be accessed for read operations contract=0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496

Running 1 test for test/MoodNftTest.t.sol:MoodNftTest
[FAIL. Reason: setup failed: the path ../img/sad.svg is not allowed to be accessed for read operations] setUp() (gas: 0)
Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 560.70µs

Running 1 test for test/DeployMoodNftTest.t.sol:DeployMoodNftTest
[PASS] testConvertSvgToUri() (gas: 26806)
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 2.09ms

Running 4 tests for test/BasicNftTest.t.sol:BasicNftTest
[PASS] testCanMintAndHaveABalance() (gas: 168902)
[PASS] testInitializedCorrectly() (gas: 14571)
[PASS] testMintWithScript() (gas: 431136)
[PASS] testTokenURIIsCorrect() (gas: 172203)
Test result: ok. 4 passed; 0 failed; 0 skipped; finished in 2.09ms
 
Ran 3 test suites: 5 tests passed, 1 failed, 0 skipped (6 total tests)

Failing tests:
Encountered 1 failing test in test/MoodNftTest.t.sol:MoodNftTest
[FAIL. Reason: setup failed: the path ../img/sad.svg is not allowed to be accessed for read operations] setUp() (gas: 0)

**Reply by dusandebelak** [reply]
I am 2 days on this XD tried all the AIs but nothing.. probably I am just stupid XD

**Reply by theirrationalone** [reply]
No no you aren't if u are then every dev is so. 😉

Please update your repo then i would be able to tell ypu what the issue is?

**Reply by dusandebelak** [reply]
Tnx man, let me push it on github and I will be happy If you take a look :) 2 min :)

**Reply by dusandebelak** [reply]
https://github.com/dusandebelak/foundry-nft-f23

It is up :) If you see anything let me know and thank you :)

**Reply by theirrationalone** [reply]
Awwwkwkwkwwkwkkk.... 

it's just a confusion almost all of the devs often face in development... 

Confusion btw Relative paths and absolute paths. [Learn more](https://stackoverflow.com/questions/21306512/difference-between-relative-path-and-absolute-path-in-javascript)

And even in start i also faced this annoying confusion... 

The confusion is that `foundry.toml::fs_permissions` args expect to be a relative path arg(s). 

And and and... 😅

`vm.readFile` expects an `absolute path`

Relative Path: accessing a path from the current directory. Ex: if you're in `script` directory then in that case you will need to access `img` folder like this using `.` member access operator proto: `../img/`

- Single dot for accessing current directory
- Double dot for accessing any parent directory


Absolute path: No matter where you're access any directory assuming you're always at the root (the first ancestor) 


`DeployMoodNft.s.sol`... 

```solidity
    function run() external returns (MoodNft) {
        if (block.chainid == 31337) {
            deployerKey = DEFAULT_ANVIL_PRIVATE_KEY;
        } else {
            deployerKey = vm.envUint("PRIVATE_KEY");
        }

        //                                                     ⏬
        string memory sadSvg = vm.readFile("../img/sad.svg");
        string memory happySvg = vm.readFile("../img/happy.svg");

        vm.startBroadcast(deployerKey);
        MoodNft moodNft = new MoodNft(svgToImageURI(sadSvg), svgToImageURI(happySvg));
        vm.stopBroadcast();
        return moodNft;
    }
```

Mitigation... 

```solidity
 //                                                                        ⏬
        string memory sadSvg = vm.readFile("img/sad.svg");
```

Don't touch `foundry.toml` file.

**Reply by dusandebelak** [reply]
Thank you, thank you, thank you  :)

now I have other errors but I think you solved for me the big one :) I had no idea. 
if you have any donate button :)
Thank you

**Reply by theirrationalone** [reply]
Your words are Enough ❤

Glad i could help ❤
